source/constructs/api/data_source/service.py

```diff
@@ -408,216 +406,219 @@ def pre_sync(jdbc: SourceJDBCConnection):
+                                         aws_secret_access_key=credentials['SecretAccessKey'],
+                                         aws_session_token=credentials['SessionToken'],
+                                         region_name=_admin_account_region)
+            """ :type : pyboto3.lakeformation """
+            # retry for grant permissions
+            num_retries = GRANT_PERMISSIONS_RETRIES
+            while num_retries > 0:
                 try:
-                    response = glue.get_crawler(Name=crawler_name)
-                    logger.info("sync_rds_connection get_crawler:")
-                    logger.info(response)
-                    try:
-                        if state == ConnectionState.ACTIVE.value or state == ConnectionState.UNSUPPORTED.value \
-                                or state == ConnectionState.ERROR.value or state == ConnectionState.STOPPING.value:
-                            up_cr_response = glue.update_crawler(
-                                Name=crawler_name,
-                                Role=crawler_role_arn,
-                                DatabaseName=glue_database_name,
-                                Targets={
-                                    'JdbcTargets': jdbc_targets,
-                                },
-                                SchemaChangePolicy={
-                                    'UpdateBehavior': 'UPDATE_IN_DATABASE',
-                                    'DeleteBehavior': 'DELETE_FROM_DATABASE'
-                                }
-                            )
-                            logger.info("update rds crawler:")
-                            logger.info(up_cr_response)
-                    except Exception as e:
-                        logger.info("update_crawler error")
-                        logger.info(str(e))
-                    st_cr_response = glue.start_crawler(
-                        Name=crawler_name
+                    response = lakeformation.grant_permissions(
+                        Principal={
+                            'DataLakePrincipalIdentifier': f"{crawler_role_arn}"
+                        },
+                        Resource={
+                            'Database': {
+                                'Name': glue_database_name
+                            }
+                        },
+                        Permissions=['ALL'],
+                        PermissionsWithGrantOption=['ALL']
                     )
-                    logger.info(st_cr_response)
                 except Exception as e:
-                    logger.info("sync_rds_connection get_crawler and create:")
-                    logger.info(str(e))
-                    response = glue.create_crawler(
+                    sleep(SLEEP_MIN_TIME)
+                    num_retries -= 1
+                else:
+                    break
+            else:
+                raise Exception('UNCONNECTED')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.